internal/providercache/installer.go

```diff
@@ -304,52 +294,36 @@ NeedProvider:
 			cb(provider, version)
 		}
 		meta, err := i.source.PackageMeta(provider, version, targetPlatform)
-		if err != nil {
-			errs[provider] = err
-			if cb := evts.FetchPackageFailure; cb != nil {
-				cb(provider, version, err)
-			}
-			continue
-		}
 
-		// if the package meta includes provider protocol versions, verify that terraform supports it.
-		if len(meta.ProtocolVersions) > 0 {
-			protoVersions := versions.MeetingConstraints(i.pluginProtocolVersion)
-			match := false
-			for _, version := range meta.ProtocolVersions {
-				if protoVersions.Has(version) {
-					match = true
-				}
+		switch err := err.(type) {
+		case nil:
+			// cool!
+		case getproviders.ErrProtocolNotSupported:
+			closestAvailable := err.Suggestion
+			if closestAvailable == &versions.Unspecified {
+				// we'll fall through to the default "not supported"
+				continue
+			}
+			var protoErrStr string
+			if version.GreaterThan(*closestAvailable) {
+				protoErrStr = providerProtocolTooNew
+			} else {
+				protoErrStr = providerProtocolTooOld
 			}
-			if match == false {
-				// Find the closest matching version
-				closestAvailable := i.findClosestProtocolCompatibleVersion(provider, version)
-				if closestAvailable == versions.Unspecified {
-					err := fmt.Errorf(errProviderVersionIncompatible, provider)
-					errs[provider] = err
-					if cb := evts.FetchPackageFailure; cb != nil {
-						cb(provider, version, err)
-					}
-					continue
-				}
 
-				// Determine if the closest matching provider is newer or older
-				// than the requirement in order to send the appropriate error
-				// message.
-				var protoErr string
-				if version.GreaterThan(closestAvailable) {
-					protoErr = providerProtocolTooNew
-				} else {
-					protoErr = providerProtocolTooOld
-				}
+			protoErr := fmt.Errorf(protoErrStr, provider, version, tfversion.String(), closestAvailable.String(), closestAvailable.String(), getproviders.VersionConstraintsString(reqs[provider]))
```

Comment:
That makes sense to me, I'll give it a go!

Reply:
Now that I look at this again, I'm not sure I understand how I could do this without a multi-line message. It's likely I am misunderstanding either your suggestion, or the "installer events" implementation.

We include a suggested version in the error message, and the version and provider name are liberally interspersed so (as it is written) I still need to customize and pass nearly that entire error message. 

Looking at the example below, I can leave out the two last lines and let `init` add them, at which point (since we already need a multi-line message) I don't see the value in doing so:

```go
const providerProtocolTooNew = `
Provider %q v%s is not compatible with Terraform %s.
You need to downgrade to v%s or earlier. Select it with
the following constraint:
	version = %q
Terraform checked all of the plugin versions matching the given constraint:
	%s
Consult the documentation for this provider for more information on
compatibility between provider and Terraform versions.
Alternatively, upgrade to the latest version of Terraform for compatibility with newer provider releases.
`
```

Can you give me any pointers? Thanks!